### PR TITLE
[GEOT-6313]: Fixes wrapping on RenderingTransformation

### PR DESCRIPTION
**Pull request by dromagnoli**
Fixes for Raster to Vector rendering transforms not respecting continuous map wrapping
## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [x] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer.

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

File: modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Mosaic.java [hunk 1]
```diff
@@ -972,15 +972,16 @@ protected RenderedImage createRenderedImage(
...
                     // no ROI, the image is full
                     RenderedImage ri = sources[i].getRenderedImage();
-                    finalROI.add(
-                            new ROIShape(
-                                    new Rectangle2D.Double(
-                                            ri.getMinX(),
-                                            ri.getMinY(),
-                                            ri.getWidth(),
-                                            ri.getHeight())));
+                    finalROI =
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Mosaic.java]: Good catch, ROI.add does not modify the current ROI, it builds a new one.

File: modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java [hunk 2]
```diff
@@ -343,29 +306,132 @@ private void addTransformedEnvelope(ReferencedEnvelope re, List<ReferencedEnvelo
...
+            // keeping into account whether we are crossing the dateline or not
+            // (halfCircles % 2)
+
+            // 3) add/subtract the original extent to get the other value of the interval
+            // in order to move the whole window (Note that the extent won't be > 360°
+            // since we are inside the "else")
+
+            int halfCircles = 0;
+            if (minX < -180) {
+                halfCircles = (int) ((Math.abs(minX) / 180));
```
Comment (aaime) [inline modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java]: Confused, any reason not to use a fullCircles and avoid the divisions later?

File: modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java [hunk 3]
```diff
@@ -343,29 +306,132 @@ private void addTransformedEnvelope(ReferencedEnvelope re, List<ReferencedEnvelo
...
+                halfCircles = (int) (Math.abs(maxX) / 180);
+                maxX += (360 * ((halfCircles / 2) + (halfCircles % 2)));
+                minX = maxX - extent;
+            } else if (maxX > 180) {
+                halfCircles = (int) (Math.abs(maxX) / 180);
+                maxX -= (360 * ((halfCircles / 2) + (halfCircles % 2)));
+                minX = maxX - extent;
+            }
+
+            if ((int) (minX / 180) < (int) (maxX / 180)) {
```
Comment (aaime) [inline modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java]: Isn't this triggering every time (maxX - minX) > 180? Doesn't mean the dateline has been crossed though.

File: modules/library/render/src/main/java/org/geotools/renderer/lite/StreamingRenderer.java [hunk 4]
```diff
@@ -2326,18 +2325,49 @@ protected GridCoverage2D readCoverage(
...
                                 throws IOException {
                             Interpolation interpolation = getRenderingInterpolation(layer);
-                            GridCoverageReaderHelper helper;
+                            RenderingHints interpolationHints =
+                                    new RenderingHints(JAI.KEY_INTERPOLATION, interpolation);
+                            final GridCoverageRenderer gcr;
                             try {
-                                helper =
-                                        new GridCoverageReaderHelper(
+                                // Use the original screenSize
```
Comment (aaime) [inline modules/library/render/src/main/java/org/geotools/renderer/lite/StreamingRenderer.java]: Uh, this is going vertical (format wise)... could you use a "convert anonymous to inner" and move this block to its own inner class?

File: modules/library/render/src/main/java/org/geotools/renderer/lite/StreamingRenderer.java [hunk 5]
```diff
@@ -2326,18 +2325,49 @@ protected GridCoverage2D readCoverage(
...
+                                // the GridCoverageRenderer will take care of eventual gutter or
```
Comment (aaime) [inline modules/library/render/src/main/java/org/geotools/renderer/lite/StreamingRenderer.java]: Is this actually true if the advanced projection handling is off though?

File: modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java [hunk 6]
```diff
@@ -343,29 +306,132 @@ private void addTransformedEnvelope(ReferencedEnvelope re, List<ReferencedEnvelo
...
         reprojectEnvelopes(sourceCRS, envelopes);
         return envelopes.stream().filter(e -> e != null).collect(Collectors.toList());
     }
 
+    /**
+     * Adjust the envelope by taking into account dateline wrapping as well as multiple spans of the
+     * whole world extent. When transform flag is true, the envelopes will be transformed before
+     * being returned
+     */
+    private void adjustEnvelope(
```
Comment (aaime) [inline modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java]: Could you add a good set of tests inside ProjectionHandlerTest covering all the possible cases handled here?

File: modules/library/render/src/main/java/org/geotools/renderer/lite/StreamingRenderer.java (hunk 5 above)
Comment (dromagnoli) [reply modules/library/render/src/main/java/org/geotools/renderer/lite/StreamingRenderer.java]: I have found another place where it's raising issues... backing to the original gridgeometry usage

File: modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java (hunk 6 above)
Comment (dromagnoli) [reply modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java]: will do.

File: modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java (hunk 2 above)
Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java]: Cool, could you add comments to explain this better then? For next time someone looks at the code.

File: modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java (hunk 3 above)
Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java]: Same as above, add comments.

File: modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java (hunk 2 above)
Comment (dromagnoli) [reply modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java]: Yep. point #2 was reporting that. I'm going to explain it better

**Comment by dromagnoli** [comment]
@aaime , added tests to ProjectionHandler to cover with the several cases
improved comments on wrapping maths and also handled your comments

**Comment by aaime** [comment]
Thanks, merging.

**Pull request merged**